batch/batch/spec_writer.py

```diff
@@ -0,0 +1,56 @@
+import secrets
+import logging
+
+log = logging.getLogger('batch.spec_writer')
+
+
+class SpecWriter:
+    byteorder = 'little'
+    signed = False
+    bytes_per_offset = 8
+
+    @staticmethod
+    def get_index_file_offsets(job_id, start_job_id):
+        assert job_id >= start_job_id
+        bytes_per_job = 2 * SpecWriter.bytes_per_offset
+        idx_start = bytes_per_job * (job_id - start_job_id)
+        idx_end = (idx_start + bytes_per_job) - 1
+        return (idx_start, idx_end)
+
+    @staticmethod
+    def read_spec_file_offsets(offsets):
+        assert len(offsets) == 2 * SpecWriter.bytes_per_offset
+        spec_start = int.from_bytes(offsets[:8], byteorder=SpecWriter.byteorder, signed=SpecWriter.signed)
+        spec_end = int.from_bytes(offsets[8:], byteorder=SpecWriter.byteorder, signed=SpecWriter.signed)
+        return (spec_start, spec_end)
+
+    def __init__(self, log_store, batch_id):
+        self.log_store = log_store
+        self.batch_id = batch_id
+        self.token = ''.join([secrets.choice('abcdefghijklmnopqrstuvwxyz0123456789') for _ in range(16)])
+
+        self._data_bytes = bytearray()
+        self._data_bytes.append(ord('['))
+
+        self._offsets_bytes = bytearray()
+        self._n_elements = 0
+
+    def add(self, data):
+        if self._n_elements > 0:
+            self._data_bytes.append(ord(','))
+
+        data_bytes = data.encode('utf-8')
+        start = len(self._data_bytes)
+        self._data_bytes.extend(data_bytes)
+        end = len(self._data_bytes) - 1
```

Comment:
I'm surprised by the -1.  I would expect end in the gs:// api not to be included in the result, so if you read from start to end, you get end - start bytes back (not end - start + 1 bytes).  Can we just confirm the behavior of the google storage client?

Reply:
I would have thought so too.

I started by looking at the source code [here](https://googleapis.dev/python/storage/latest/_modules/google/cloud/storage/blob.html#Blob.download_as_string) and tracking the underlying method calls to `_do_download` on that same page (can't make a link to it).

I tracked it further to the actual download class [here](https://googleapis.github.io/google-resumable-media-python/latest/google.resumable_media.requests.html#simple-downloads) and saw this example just to confirm `_do_download`'s docstring was correct:

```
>>> download = Download(media_url, start=4096, end=8191)
>>> response = download.consume(transport)
>>> download.finished
True
>>> response
<Response [206]>
>>> response.headers[u'Content-Length']
'4096'
>>> response.headers[u'Content-Range']
'bytes 4096-8191/1364156'
>>> len(response.content)
4096
```

```
        :type start: int
        :param start: Optional, the first byte in a range to be downloaded.

        :type end: int
        :param end: Optional, The last byte in a range to be downloaded.
        """
 
```

Reply:
Surprising!  Thanks for confirming.  This might be a world a comment in some relevant place that acknowledging this is surprising and that it has been confirmed.